src/library/blast/arith/heuristic.cpp

```diff
@@ -0,0 +1,552 @@
+/*
+Copyright (c) 2015 Daniel Selsam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Daniel Selsam
+*/
+#include <vector>
+#include <algorithm>
+#include <memory>
+#include "kernel/abstract.h"
+#include "util/sexpr/option_declarations.h"
+#include "util/numerics/mpq.h"
+#include "library/constants.h"
+#include "library/blast/proof_expr.h"
+#include "library/blast/arith/heuristic.h"
+#include "library/blast/arith/num.h"
+#include "library/blast/arith/polynomial.h"
+#include "library/blast/arith/normalize.h"
+#include "library/blast/arith/normalize_poly.h"
+#include "library/blast/backward/backward_strategy.h"
+#include "library/blast/blast.h"
+#include "library/blast/trace.h"
+#include "library/blast/options.h"
+#include "library/blast/simplifier/simplifier.h"
+#include "library/num.h"
+
+#ifndef LEAN_DEFAULT_HEURISTIC_MAX_STEPS_PER_ACTION
+#define LEAN_DEFAULT_HEURISTIC_MAX_STEPS_PER_ACTION 10
+#endif
+
+namespace lean {
+namespace blast {
+
+/** Globals */
+static name * g_heuristic_trace_name           = nullptr;
+static name * g_heuristic_max_steps_per_action = nullptr;
+
+static unsigned g_ext_id = 0;
+
+struct arith_heuristic_proof_step_cell : public proof_step_cell {
+    list<expr>     m_placeholders;     // placeholders for non-zero proofs
+
+    virtual ~arith_heuristic_proof_step_cell() {}
+    arith_heuristic_proof_step_cell(list<expr> const & placeholders):
+        m_placeholders(placeholders) {}
+
+    virtual action_result resolve(expr const & pf) const override {
+        buffer<expr> proofs;
+        for (expr const & placeholder : m_placeholders) {
+            optional<expr> pf;
+            {
+                flet<state> save_state(curr_state(), curr_state());
+                curr_state().set_target(mlocal_type(placeholder));
+                pf = mk_backward_strategy()();
```

Comment:
If we do keep proofs with applications of `inv`, this issue will no longer affect `resolve`. Either way, we are going to want an extensible strategy specific to proving `! (_ = 0)`, though the strategy we choose will depend on whether we keep proofs with `inv`, since in the latter case we will want to be more robust at the possible expense of efficiency.